### PR TITLE
add cookbook for mkl binary classification

### PR DESCRIPTION
**Pull request by sanuj**
I have converted the code in `examples/undocumented/python_modular/mkl_binclass_modular.py` to meta language for now. I think it contains some redundancies. Need feedback on that.

### CODE REVIEW

**Comment by karlnapf** [comment]
travis fails?

**Comment by karlnapf** [comment]
no integration testing data yet

File: doc/cookbook/source/examples/classifier/mkl_binary_class.rst [hunk 1]
```diff
@@ -0,0 +1,59 @@
+==================================================
+Multiple Kernel Learning for Binary Classification
+==================================================
+
+Multiple kernel learning (MKL) is about using a combined kernel i.e. a kernel consisting of a linear combination of arbitrary kernels over different domains.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: avoid "is about", but rather say "MKL is based on convex combinations of ..."

File: doc/cookbook/source/examples/classifier/mkl_binary_class.rst [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+.. math::
+
+    f({\bf x})=\text{sign} \left(\sum_{i=0}^{N-1} \alpha_i k({\bf x}, {\bf x_i})+b\right)
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: This is for classification, you should say that "For example, for a kernel SVM the model is"

Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: Actually, I would not give this line here, this is already mentioned in the kernel svm cookbook (you could reference that like @OXPHOS did in #3242 ).
Rather the math line should be the k=\sum_i k_i .....

File: doc/cookbook/source/references.bib [hunk 3]
```diff
@@ -59,4 +59,13 @@ @book{scholkopf2002learning
   author={B. Sch{\"o}lkopf and A.J. Smola},
   year={2002},
   publisher={MIT press}
+}
+@article{sonnenburg2006large,
```
Comment (karlnapf) [inline doc/cookbook/source/references.bib]: Are the algorithms in Shogun the ones in this paper?

File: doc/cookbook/source/examples/classifier/mkl_binary_class.rst [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+One could make a combination of kernels like:
+
+.. math::
+
+    {\bf k}(x_i,x_j)=\sum_{k=0}^{K} \beta_k {\bf k_k}(x_i, x_j)
+
+where :math:`\beta_k > 0` and   :math:`\sum_{k=0}^{K} \beta_k = 1`
+
+Optimal :math:`\alpha_i`, :math:`\beta` and :math:`b` are determined via training.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: I think this cookbook should solely focus on the MKL part
 * settup up multiple kernels
 * running mkl
 * extracting kernel weights

And refer to the kernel svm cookbook for all other things

File: doc/cookbook/source/examples/classifier/mkl_binary_class.rst [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+-------
+Example
+-------
+
+Imagine we have files with training and test data. We create CDenseFeatures (here 64 bit floats aka RealFeatures) and :sgclass:`CBinaryLabels` as
+
+.. sgexample:: mkl_binary_class.sg:create_features
+
+Then we create an instance of :sgclass:`CPolyKernel`, initialize it with training data and extract a kernel matrix.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: The point of this is to show that one can combine combined kernels with other ones. This doesnt really shine in this way. Maybe start by saying that one can combine all shogun kernels, and that you will combine a custom, a gaussian, and a poly.

File: doc/cookbook/source/examples/classifier/mkl_binary_class.rst [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+We create an instance of :sgclass:`CCombinedKernel` by using :sgclass:`CCustomKernel` and :sgclass:`CPolyKernel`. In order to initialize :sgclass:`CCombineKernel` with features, we also need to create an object of :sgclass:`CCombinedFeatures`.
+
+.. sgexample:: mkl_binary_class.sg:create_combined_train
+
+We create an object of :sgclass:`CMKLClassification`, provide the combined kernel and labels before training it.
+
+.. sgexample:: mkl_binary_class.sg:train_mkl
+
+As shown above for training data, we similarly make a :sgclass:`CCombinedKernel` for testing data.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: why a new kernel for testing?

Comment (sanuj) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: If I want to use the same `CombinedKernel` then I would need to change the kernel matrix for `CustomKernel` by using `set_full_kernel_matrix_from_full` which is not possible with meta language as `get_kernel()` returns `Kernel*` and not `CustomKernel*`:
```
CustomKernel k = combined_kernel.get_kernel(0)
k.set_full_kernel_matrix_from_full(poly_kernel_mat_test)
```
Other way is to delete the old `CustomKernel` and append a new one. @karlnapf do you want me to do this?

Comment (sanuj) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: @karlnapf or i can also remove `CustomKernel` and only keep `PolyKernel` and `GaussianKernel`. This way I can just reuse the old `CombinedKernel` without any problems.

File: doc/cookbook/source/references.bib (hunk 3 above)
Comment (sanuj) [inline doc/cookbook/source/references.bib]: @karlnapf This paper is cited in `MKL.h` documentation.

File: doc/cookbook/source/examples/classifier/mkl_binary_class.rst [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+Multiple Kernel Learning for Binary Classification
+==================================================
+
+Multiple kernel learning (MKL) is based on convex combinations of arbitrary kernels over different domains. One could make a combination of kernels like:
+
+.. math::
+
+    {\bf k}(x_i,x_j)=\sum_{i=1}^{K} \beta_k {\bf k}_i(x_i, x_j)
+
+where :math:`\beta_k > 0`, :math:`\sum_{k=1}^{K} \beta_k = 1`, :math:`K` is the number of sub-kernels, :math:`\bf{k}` is a combined kernel matrix, :math:`{\bf k}_i` is an individual kernel matrix and :math:`x_i` is a data point.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: the k guys are not kernel matrices, but just kernels

File: doc/cookbook/source/examples/classifier/mkl_binary_class.rst [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+Classification is done by using Support Vector Machines (SVM). `See the linear SVM cookbook <http://shogun.ml/cookbook/latest/examples/classifier/linear_svm.html>`_ for more details. Optimal :math:`\alpha` and :math:`b` for SVM and :math:`\beta` are determined via training.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: I think we should not mention the linear SVM coobkook here, but a LibSVM (kernel svm) cookbook

File: doc/cookbook/source/examples/classifier/mkl_binary_class.rst (hunk 7 above)
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: and \{x_i\}_i are the training data

File: doc/cookbook/source/examples/classifier/mkl_binary_class.rst [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+Imagine we have files with training and test data. We create CDenseFeatures (here 64 bit floats aka RealFeatures) and :sgclass:`CBinaryLabels` as
+
+.. sgexample:: mkl_binary_class.sg:create_features
+
+Then we create indvidual kernels like :sgclass:`CPolyKernel`, :sgclass:`CCustomKernel` and :sgclass:`CGaussianKernel` which will be later combined in one :sgclass:`CCombinedKernel`.
+
+.. sgexample:: mkl_binary_class.sg:create_kernel
+
+We create an instance of :sgclass:`CCombinedKernel` and append the :sgclass:`CKernel` objects. In order to initialize :sgclass:`CCombineKernel` with features, we also need to create an object of :sgclass:`CCombinedFeatures`.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: The combined features are for the case where different kernels have different features.
You dont need to do that here, but mention it. Also you combined features only get two copies of the training features here. Is that because one kernel is precomputed?

File: doc/cookbook/source/examples/classifier/mkl_binary_class.rst [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+.. sgexample:: mkl_binary_class.sg:create_combined_train
+
+We create an object of :sgclass:`CMKLClassification`, provide the combined kernel and labels before training it.
+
+.. sgexample:: mkl_binary_class.sg:train_mkl
+
+After training, we can extract :math:`\beta`, :math:`\alpha` and :math:`b`.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: Maybe say "the SVM coefficients \alpha"

File: doc/cookbook/source/examples/classifier/mkl_binary_class.rst [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+-------
+Example
+-------
+
+Imagine we have files with training and test data. We create CDenseFeatures (here 64 bit floats aka RealFeatures) and :sgclass:`CBinaryLabels` as
+
+.. sgexample:: mkl_binary_class.sg:create_features
+
+Then we create indvidual kernels like :sgclass:`CPolyKernel`, :sgclass:`CCustomKernel` and :sgclass:`CGaussianKernel` which will be later combined in one :sgclass:`CCombinedKernel`.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: put (precomputed) before CCustomKernel

File: doc/cookbook/source/examples/classifier/mkl_binary_class.rst [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+We create an object of :sgclass:`CMKLClassification`, provide the combined kernel and labels before training it.
+
+.. sgexample:: mkl_binary_class.sg:train_mkl
+
+After training, we can extract :math:`\beta`, :math:`\alpha` and :math:`b`.
+
+.. sgexample:: mkl_binary_class.sg:extract_weights
+
+We update the :sgclass:`CCombinedKernel` object for testing data. We also need to update the :sgclass:`CCustomKernel` object because it is not done by :sgclass:`CCombinedKernel` directly.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: "because it was precomputed"

File: doc/cookbook/source/examples/classifier/mkl_binary_class.rst [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+.. sgexample:: mkl_binary_class.sg:create_combined_test
+
+Finally, we set the updated kernel and predict :sgclass:`CBinaryLabels` for test data.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: Can you also compute accuracy (and make sure it looks reasonable)

File: doc/cookbook/source/examples/classifier/mkl_binary_class.rst [hunk 14]
```diff
@@ -0,0 +1,55 @@
+==================================================
+Multiple Kernel Learning for Binary Classification
+==================================================
+
+Multiple kernel learning (MKL) is based on convex combinations of arbitrary kernels over different domains. One could make a combination of kernels like:
+
+.. math::
+
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: I wonder whether we should say one sentence about how the MKL solver works here.

File: doc/cookbook/source/examples/classifier/mkl_binary_class.rst [hunk 15]
```diff
@@ -0,0 +1,55 @@
+==================================================
+Multiple Kernel Learning for Binary Classification
+==================================================
+
+Multiple kernel learning (MKL) is based on convex combinations of arbitrary kernels over different domains. One could make a combination of kernels like:
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: make it "(over potentially different domains)"

Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: Remove "One could make a combination of kernels like:"

**Comment by karlnapf** [comment]
One more iteration and this is done. Looks nice already! :)

**Comment by karlnapf** [comment]
Ah travis moans, checking

File: doc/cookbook/source/examples/classifier/mkl_binary_class.rst [hunk 16]
```diff
@@ -0,0 +1,55 @@
+==================================================
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: can you avoid abbreviations in the filenames. Use multiple_kernel_learning_binary_classifier.rst

File: doc/cookbook/source/examples/classifier/multiclass_linearmachine.rst [hunk 17]
```diff
@@ -6,7 +6,7 @@ We extend the application of linear machines to multi-class datasets by construc
 
 In this example, we show how to apply :sgclass:`CLibLinear` to multi-class cases with :sgclass:`CLinearMulticlassMachine`.
 
-`See the linear SVM cookbook <http://shogun.ml/cookbook/latest/examples/classifier/linear_svm.html>`_ for the infomration about :sgclass:`CLibLinear` binary classifier.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_linearmachine.rst]: please send a separate patch for this. But thanks and well spotted

File: doc/cookbook/source/examples/classifier/mkl_binary_class.rst (hunk 8 above)
Comment (sanuj) [inline doc/cookbook/source/examples/classifier/mkl_binary_class.rst]: @karlnapf i just started a PR for kernel svm #3281. We can change this once this gets merged.

File: src/shogun/kernel/CombinedKernel.cpp [hunk 18]
```diff
@@ -78,8 +79,12 @@ bool CCombinedKernel::init(CFeatures* l, CFeatures* r)
 		CCombinedFeatures* combined_r=new CCombinedFeatures();
 		for (index_t i=0; i<get_num_subkernels(); ++i)
 		{
-			combined_l->append_feature_obj(l);
-			combined_r->append_feature_obj(r);
+			k = get_kernel(i);
```
Comment (sanuj) [inline src/shogun/kernel/CombinedKernel.cpp]: @karlnapf if you find this appropriate, i can send a separate patch for this.

File: doc/cookbook/source/examples/classifier/multiple_kernel_learning_binary_classifier.rst [hunk 19]
```diff
@@ -0,0 +1,59 @@
+==================================================
+Multiple Kernel Learning for Binary Classification
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiple_kernel_learning_binary_classifier.rst]: Can you remove "for binary classification"

File: src/shogun/kernel/CombinedKernel.cpp (hunk 18 above)
Comment (karlnapf) [inline src/shogun/kernel/CombinedKernel.cpp]: yeah, different patch definitely. Always try to separate things.

File: src/shogun/kernel/CombinedKernel.cpp [hunk 20]
```diff
@@ -98,7 +103,7 @@ bool CCombinedKernel::init(CFeatures* l, CFeatures* r)
 
 	CFeatures* lf=NULL;
 	CFeatures* rf=NULL;
-	CKernel* k=NULL;
+	// CKernel* k=NULL;
```
Comment (karlnapf) [inline src/shogun/kernel/CombinedKernel.cpp]: remove

File: src/shogun/kernel/CombinedKernel.cpp (hunk 18 above)
Comment (karlnapf) [inline src/shogun/kernel/CombinedKernel.cpp]: I think this might increase the ref counter, you have to unref later

**Comment by sanuj** [comment]
@karlnapf can't understand why is Ruby failing here.

**Comment by karlnapf** [comment]
Compile ruby and run it locally!
Ctest -R testname

**Comment by sanuj** [comment]
@karlnapf Getting [this](https://gist.github.com/sanuj/cc98f73b040b34f8adbc1da12af71aa6) error while building.

**Comment by karlnapf** [comment]
I don't think this is error is related to the example. Look, it is caused by `ld`.
Try to isolate the problem, and then ping me in irc

**Comment by sorig** [comment]
This Ruby error message is cryptic. You need to install the Ruby modular interface, take the generated ruby file and figure why it doesn't work. To me it doesn't look like it's a meta-language issue, but rather a problem with the SWIG interface.

Let me know if it turns out to be a meta-language issue. In that case you may need to update the `targets/ruby.json` file.

**Comment by karlnapf** [comment]
I am merging this as travis (apart from windows) was actually fine.

**Comment by karlnapf** [comment]
Needs data
